File: puzzle-generator/generator.py
import random
import json
import sys
from wordfreq import word_frequency
from datetime import date

# ── Configuration ────────────────────────────────────────────────────────────

import os
WORDLIST_PATH = os.path.join(os.path.dirname(__file__), "enable1.txt")
MIN_WORDS = 20
MAX_WORDS = 80
MIN_WORD_LENGTH = 4
FREQUENCY_THRESHOLD = 3e-7  # words below this are considered too obscure

PROFANITY_BLOCKLIST = {
    "fuck", "shit", "cunt", "cock", "dick", "pussy", "ass", "bitch",
    "nigger", "nigga", "faggot", "fag", "slut", "whore", "bastard",
    "piss", "tits", "twat", "wank", "spic", "kike", "chink"
}

# ── Load word list ────────────────────────────────────────────────────────────

def load_wordlist(path):
    with open(path) as f:
        words = [w.strip().lower() for w in f if w.strip()]
    # Filter to words that are clean, common enough, and long enough
    return [
        w for w in words
        if len(w) >= MIN_WORD_LENGTH
        and w not in PROFANITY_BLOCKLIST
        and word_frequency(w, "en") >= FREQUENCY_THRESHOLD
    ]

# ── Puzzle validation ─────────────────────────────────────────────────────────

def is_valid_word(word, center, letters):
    """Check if a word qualifies for a given puzzle."""
    letter_set = set(letters)
    return (
        len(word) >= MIN_WORD_LENGTH
        and center in word
        and all(c in letter_set for c in word)
    )

def is_pangram(word, letters):
    """Check if a word uses all 7 puzzle letters at least once."""
    return set(letters).issubset(set(word))

def score_word(word, letters):
    """Calculate points for a word using NYT rules."""
    if len(word) == 4:
        points = 1
    else:
        points = len(word)
    if is_pangram(word, letters):
        points += 7
    return points

# ── Puzzle generation ─────────────────────────────────────────────────────────

def find_best_center(pangram_letters, wordlist):
    """
    Choose the center letter that produces the most valid words.
    """
    best_center = None
    best_count = 0
    for candidate in pangram_letters:
        count = sum(
            1 for w in wordlist
            if is_valid_word(w, candidate, pangram_letters)
        )
        if count > best_count:
            best_count = count
            best_center = candidate
    return best_center, best_count

def generate_puzzle(seed=None):
    """
    Generate a daily puzzle. Returns a dict with:
      - letters: list of 7 letters (center first)
      - center: the center letter
      - valid_words: sorted list of valid words
      - pangrams: list of pangrams in the valid word list
      - word_count: number of valid words
      - max_score: maximum possible score
    """
    if seed is None:
        seed = date.today().isoformat()

    print(f"Loading word list...", file=sys.stderr)
    wordlist = load_wordlist(WORDLIST_PATH)
    print(f"Loaded {len(wordlist)} usable words after filtering.", file=sys.stderr)

    # Find all pangram candidates (words with exactly 7 unique letters)
    pangram_candidates = [w for w in wordlist if len(set(w)) == 7]
    print(f"Found {len(pangram_candidates)} pangram candidates.", file=sys.stderr)

    rng = random.Random(seed)
    rng.shuffle(pangram_candidates)

    for pangram in pangram_candidates:
        letters = list(set(pangram))  # exactly 7 unique letters

        center, word_count = find_best_center(letters, wordlist)

        if word_count < MIN_WORDS:
            continue

        # Build the valid word list
        valid_words = sorted([
            w for w in wordlist
            if is_valid_word(w, center, letters)
        ])

        if word_count > MAX_WORDS:
            continue

        # Success — build the result
        pangrams_found = [w for w in valid_words if is_pangram(w, letters)]
        max_score = sum(score_word(w, letters) for w in valid_words)

        # Put center letter first in the letters list
        other_letters = [l for l in letters if l != center]
        return {
            "seed": seed,
            "center": center,
            "letters": [center] + sorted(other_letters),
            "valid_words": valid_words,
            "pangrams": pangrams_found,
            "word_count": word_count,
            "max_score": max_score,
        }

    return None  # Should rarely happen

# ── Main ──────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    args = sys.argv[1:]
    json_mode = "--json" in args
    date_args = [a for a in args if a != "--json"]
    seed = date_args[0] if date_args else None

    puzzle = generate_puzzle(seed=seed)

    if puzzle is None:
        print(json.dumps({"error": "Failed to generate a valid puzzle"}))
        sys.exit(1)
    else:
        if json_mode:
            print(json.dumps(puzzle))
        else:
            print(f"\n{'='*50}")
            print(f"PUZZLE FOR {puzzle['seed']}")
            print(f"{'='*50}")
            print(f"Letters:   {' '.join(l.upper() for l in puzzle['letters'])}")
            print(f"Center:    {puzzle['center'].upper()}")
            print(f"Words:     {puzzle['word_count']}")
            print(f"Pangrams:  {', '.join(puzzle['pangrams'])}")
            print(f"Max score: {puzzle['max_score']}")
            print(f"\nValid words:")
            for i, word in enumerate(puzzle['valid_words'], 1):
                marker = " ★" if word in puzzle['pangrams'] else ""
                print(f"  {i:3}. {word}{marker}")

File: server/generator.js
// Copyright (c) 2026 Page Stites
// All rights reserved.

const fs = require("fs");
const path = require("path");

// ── Configuration ────────────────────────────────────────────────────────────

const WORDLIST_PATH = path.join(__dirname, "../puzzle-generator/enable1.txt");
const MIN_WORDS = 20;
const MAX_WORDS = 80;
const MIN_WORD_LENGTH = 4;

const PROFANITY_BLOCKLIST = new Set([
  "fuck", "shit", "cunt", "cock", "dick", "pussy", "ass", "bitch",
  "nigger", "nigga", "faggot", "fag", "slut", "whore", "bastard",
  "piss", "tits", "twat", "wank", "spic", "kike", "chink"
]);

// Common English words frequency approximation
// We'll use word length and pattern as a proxy since we don't have wordfreq in JS
// Words shorter than 4 letters are already filtered; we filter very rare long words
// by checking against a minimum character frequency pattern
function isCommonEnough(word) {
  // Filter out words with rare letter combinations that suggest obscurity
  // Words with Q not followed by U, words with 3+ consecutive consonants, etc.
  const consonants = "bcdfghjklmnpqrstvwxyz";
  let consecutiveConsonants = 0;
  let maxConsecutive = 0;
  for (const ch of word) {
    if (consonants.includes(ch)) {
      consecutiveConsonants++;
      maxConsecutive = Math.max(maxConsecutive, consecutiveConsonants);
    } else {
      consecutiveConsonants = 0;
    }
  }
  if (maxConsecutive >= 4) return false;
  if (word.includes("q") && !word.includes("qu")) return false;
  if (word.length > 12) return false;
  return true;
}

// ── Load word list ────────────────────────────────────────────────────────────

let cachedWordlist = null;

function loadWordlist() {
  if (cachedWordlist) return cachedWordlist;
  const raw = fs.readFileSync(WORDLIST_PATH, "utf8");
  const words = raw.split("\n").map(w => w.trim().toLowerCase()).filter(w => w);
  cachedWordlist = words.filter(w =>
    w.length >= MIN_WORD_LENGTH &&
    !PROFANITY_BLOCKLIST.has(w) &&
    isCommonEnough(w) &&
    /^[a-z]+$/.test(w)
  );
  console.error(`Loaded ${cachedWordlist.length} words after filtering.`);
  return cachedWordlist;
}

// ── Puzzle logic ──────────────────────────────────────────────────────────────

function isValidWord(word, center, letterSet) {
  if (word.length < MIN_WORD_LENGTH) return false;
  if (!word.includes(center)) return false;
  for (const ch of word) {
    if (!letterSet.has(ch)) return false;
  }
  return true;
}

function isPangram(word, letters) {
  return letters.every(l => word.includes(l));
}

function scoreWord(word, letters) {
  let points = word.length === 4 ? 1 : word.length;
  if (isPangram(word, letters)) points += 7;
  return points;
}

function seededRandom(seed) {
  // Simple seeded PRNG (mulberry32)
  let h = 0;
  for (let i = 0; i < seed.length; i++) {
    h = Math.imul(31, h) + seed.charCodeAt(i) | 0;
  }
  return function() {
    h |= 0; h = h + 0x6D2B79F5 | 0;
    let t = Math.imul(h ^ h >>> 15, 1 | h);
    t = t + Math.imul(t ^ t >>> 7, 61 | t) ^ t;
    return ((t ^ t >>> 14) >>> 0) / 4294967296;
  };
}

function shuffle(arr, rng) {
  const a = [...arr];
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

// ── Main generator ────────────────────────────────────────────────────────────

function generatePuzzle(seed) {
  if (!seed) seed = new Date().toISOString().slice(0, 10);

  const wordlist = loadWordlist();

  // Find pangram candidates (words with exactly 7 unique letters)
  const candidates = wordlist.filter(w => new Set(w).size === 7);
  console.error(`Found ${candidates.length} pangram candidates.`);

  const rng = seededRandom(seed);
  const shuffled = shuffle(candidates, rng);

  for (const pangram of shuffled) {
    const letters = [...new Set(pangram)]; // exactly 7 unique letters

    // Find best center letter
    let bestCenter = null;
    let bestCount = 0;
    for (const candidate of letters) {
      const letterSet = new Set(letters);
      const count = wordlist.filter(w => isValidWord(w, candidate, letterSet)).length;
      if (count > bestCount) {
        bestCount = count;
        bestCenter = candidate;
      }
    }

    if (bestCount < MIN_WORDS) continue;

    const letterSet = new Set(letters);
    const validWords = wordlist
      .filter(w => isValidWord(w, bestCenter, letterSet))
      .sort();

    if (validWords.length > MAX_WORDS) continue;

    const pangrams = validWords.filter(w => isPangram(w, letters));
    const maxScore = validWords.reduce((sum, w) => sum + scoreWord(w, letters), 0);
    const otherLetters = letters.filter(l => l !== bestCenter).sort();

    return {
      seed,
      center: bestCenter,
      letters: [bestCenter, ...otherLetters],
      valid_words: validWords,
      pangrams,
      word_count: validWords.length,
      max_score: maxScore,
    };
  }

  return null;
}

module.exports = { generatePuzzle };